app/lib/analyzer/task_sources.dart

```diff
@@ -5,44 +5,52 @@
 import 'dart:async';
 
 import 'package:gcloud/db.dart';
+import 'package:logging/logging.dart';
 
 import '../frontend/models.dart';
 import '../shared/task_scheduler.dart';
 
 import 'models.dart';
 
+final Logger _logger = new Logger('pub.analyzer.source');
+
 /// Creates a task when a version uploaded in the past 10 minutes has no
 /// analysis yet.
-class DatastoreHeadTaskSource extends PollingTaskSource {
+class DatastoreHeadTaskSource implements TaskSource {
   final DatastoreDB _db;
   DateTime _lastTs;
-  DatastoreHeadTaskSource(this._db) : super(const Duration(minutes: 1));
+  DatastoreHeadTaskSource(this._db);
 
   @override
-  Future poll() async {
-    final DateTime now = new DateTime.now().toUtc();
-    final DateTime tenMinutesAgo = now.subtract(const Duration(minutes: 10));
-    DateTime minCreated;
-    if (_lastTs == null) {
-      minCreated = tenMinutesAgo;
-    } else if (_lastTs.isBefore(tenMinutesAgo)) {
-      // more than ten minutes passed since the last poll
-      minCreated = _lastTs;
-    } else {
-      minCreated = tenMinutesAgo;
-    }
-    _lastTs = tenMinutesAgo;
+  Stream<Task> startStreaming() async* {
+    for (;;) {
+      try {
+        final DateTime now = new DateTime.now().toUtc();
+        final DateTime tenMinutesAgo =
+            now.subtract(const Duration(minutes: 10));
+        _lastTs ??= tenMinutesAgo;
+        final DateTime minCreated =
+            _lastTs.isBefore(tenMinutesAgo) ? _lastTs : tenMinutesAgo;
 
-    final Query q = _db.query(PackageVersion)..filter('created >=', minCreated);
-    await for (PackageVersion pv in q.run()) {
-      final List<PackageVersionAnalysis> items = await _db.lookup([
-        _db.emptyKey
-            .append(PackageAnalysis, id: pv.package)
-            .append(PackageVersionAnalysis, id: pv.version)
-      ]);
-      if (items.first == null) {
-        addTask(new Task(pv.package, pv.version));
+        final Query q = _db.query(PackageVersion)
+          ..filter('created >=', minCreated);
```

Comment:
You could add `..order('created')` here, to ensure if an exception occurs during querying/analysis, the update of `_lastTs` is strictly increasing (i.e. we don't start from newest to oldest, but from oldest to newest).